### DISCUSSION TITLE
When to use nested ProviderScopes?

### DISCUSSION THREAD

**Discussion by richardjharris** [rrousselGit/riverpod#1130] [Q&A]
Hello, I'm a bit confused about using scoped providers as explained here: https://codewithandrea.com/articles/flutter-state-management-riverpod/  specifically "This works but if the ListView rebuilds, all of its children will rebuild too."

I thought that if that if you use key = product_id, a ListView will not rebuild items with the same key, even if the list changes? Am I wrong? And also, how does making the `ProductItem()` const work? I guess I don't understand how the class can be const but _also_ rebuild individual items when those products change, unless riverpod does some kind of internal caching.

Finally I find the whole providerscope/override/const object kind of ugly. For normal use-cases, is it okay to just use `ref.watch(someStreamProvider)` and then build `ProductItem(item: item)` for each item?

Loving the library, just a bit confused about this topic. Thanks a lot.

Rich

**Comment by rrousselGit** [comment]
> I thought that if that if you use key = product_id, a ListView will not rebuild items with the same key, even if the list changes? Am I wrong? 

Keys don't do that. They are to help Flutter determine if the widget in which items are displayed changed, not to avoid rebuild.


> Finally I find the whole providerscope/override/const object kind of ugly. For normal use-cases, is it okay to just use ref.watch(someStreamProvider) and then build ProductItem(item: item) for each item?

You do you, but as mentioned, whenever something changes, all the items in the list will update
